[Anant Shanker] iP

Anant1902 opened the pull request:
# DrakePro 😎 

> "Soon is not as good as now.” – Seth Godin ([Source](https://dansilvestre.com/productivity-quotes/#9_%E2%80%9CSoon_is_not_as_good_as_now%E2%80%9D_%E2%80%93_Seth_Godin))

_Don't just listen to Drake, let him manage your tasks_ - in the most efficient way possible.

- CLI first
- newbie friendly
- ~~fast~~ INCREDIBLY fast

To begin now:
1. Clone [this](https://github.com/Anant1902/ip) Github Repository
2. Run it on your local machine using Java 11
3. let the magic begin

and its **COMPLETELY FREE**

Features:

- [ ] Manage Todos
- [ ] Manage Deadlines
- [ ] Manage Events
- [ ] and many more coming soon...

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) throws Exception {
        new Drake("./list.dat").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Anant1902 committed:
Add greeting skeleton

Anant1902 committed:
Add echo feature

Anant1902 committed:
Enable add and list feature

Anant1902 committed:
Add mark feature

Anant1902 committed:
add todo, events, deadlines

Anant1902 committed:
Test using i/o redirection

Anant1902 committed:
Handle erros

Anant1902 committed:
add delete function

Anant1902 committed:
Use enums

Anant1902 committed:
Enabled file save

Anant1902 committed:
Merge branch 'branch-Level-7'

Anant1902 committed:
Enable date recognition

Anant1902 committed:
Merge branch 'branch-Level-8'

Anant1902 committed:
Add OOP

Anant1902 committed:
Structure into packages

Anant1902 committed:
Merge branch 'add-gradle-support'

Anant1902 committed:
add gradle support

Anant1902 committed:
Add 2 junit tests

Anant1902 committed:
Add javadocs

Anant1902 committed:
Follow coding standard

Anant1902 committed:
Add find feat

Anant1902 committed:
Fix errors in coding standard

Anant1902 committed:
Merge branch 'branch-A-javaDoc' into branch-A-CodingStandard

Anant1902 committed:
Merge branch 'branch-Level-9' into branch-A-CodingStandard

Anant1902 committed:
Merge branch 'branch-A-CodingStandard'

Anant1902 committed:
Configure checkstyle

Anant1902 committed:
Fix UI with checkstyle

Anant1902 committed:
Fix checkstyle for all

Anant1902 committed:
Add working GUI

Anant1902 committed:
Add images

Anant1902 committed:
Merge branch-level-10 into master

Shauryan123 commented on src/main/java/drake/Drake.java:
```diff
@@ -0,0 +1,197 @@
+        }
+    }
+}
+
```
Good use of enums here, good organisation of code through switch-case statements to increase performance

zengzihui replied:
I like how you use enums here to represent different commands, it helps to improve readability and maintainability.

Shauryan123 commented on src/main/java/drake/Main.java:
```diff
@@ -0,0 +1,33 @@
+            stage.show();
+            fxmlLoader.<MainWindow>getController().setWelcomeMessage(Ui.showWelcome());
+        } catch (IOException e) {
+            e.printStackTrace();
```
Consider handling the IOException in a better way . Maybe show a user-friendly error message or log the error with more context.

Shauryan123 commented on src/main/java/drake/Drake.java:
```diff
@@ -0,0 +1,197 @@
+    private TextField userInput;
+    private Button sendButton;
+    private Scene scene;
+    private Image user = new Image(this.getClass().getResourceAsStream("/images/user.jpeg"));
+    private Image duke = new Image(this.getClass().getResourceAsStream("/images/drake.jpeg"));
```
```suggestion
    private static final String USER_IMAGE_PATH = "/images/user.jpeg";
    private static final String DRAKE_IMAGE_PATH = "/images/drake.jpeg";

    private Image user = new Image(this.getClass().getResourceAsStream(USER_IMAGE_PATH));
    private Image duke = new Image(this.getClass().getResourceAsStream(DRAKE_IMAGE_PATH));
```

Shauryan123 replied:
Avoid using magic strings directly in the code. Use named constants for file paths and resource paths to improve maintainability.

Shauryan123 commented on src/main/java/drake/Ui.java:
```diff
@@ -0,0 +1,125 @@
+     */
+    public String showTaskList(TaskList tasks) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("____________________________________________________________\n");
```
Try abstracting out the divider line code that is reused in multiple functions like showAddTask, showTaskList etc to a seperate divider function. this make the code clumsy

Shauryan123 replied:
```suggestion
     public void divider() {
        System.out.println("____________________________________________________________\n");
        
        }
```

Shauryan123 commented on src/main/java/drake/task/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+     * @param index The index of the task to be marked as done.
+     * @throws IndexOutOfBoundsException If the index is out of the range of the task list.
+     */
+    public void markTask(int index) throws IndexOutOfBoundsException {
+        getTask(index).isDone = true;
+    }
+
+    /**
+     * Un-marks a task in the task list, indicating it is not done.
+     *
+     * @param index The index of the task to be unmarked.
+     * @throws IndexOutOfBoundsException If the index is out of the range of the task list.
+     */
+    public void unmarkTask(int index) throws IndexOutOfBoundsException {
+        getTask(index).isDone = false;
+    }
```
```suggestion
public void markTask(int index) {
    Task task = getTask(index); // Ensures index bounds are checked.
    task.markAsDone(); // Assuming Task has a markAsDone method.
}

public void unmarkTask(int index) {
    Task task = getTask(index); // Ensures index bounds are checked.
    task.unmarkAsDone(); // Assuming Task has an unmarkAsDone method.
}

```

Shauryan123 replied:
Direct access to Task's isDone field violates encapsulation principles. Assuming Task is another class within the drake.task package, it should provide methods to encapsulate its state changes

Shauryan123 reviewed (commented):
I think overall, the code is well orgnaised and complies with most coding stadnards. Ata  few places, more functional abstraction could be done as certain sections of code are reused, styling and indentation was correctly followed

Anant1902 committed:
Update build.gradle

The current JavaFX UI has to be run from terminal.

Compile the libraries and the code into a .jar file.

This simplifies the program UI into a one-click application.

Anant1902 committed:
Update src/main/java/drake/Drake.java

Co-authored-by: Shauryan123 <119833428+Shauryan123@users.noreply.github.com>

Anant1902 committed:
Add assertions

Functions do not handle assertions that are necessary to hold.

Asserting input for relevant functions is necessary to avoid invalid arguments.

'assert' is the simplest and most straightforward way of achieving this.

Anant1902 committed:
Merge pull request #1 from Anant1902/branch-A-Assertions

Add assertions

Anant1902 committed:
Refactor codebase.

Codebase has repetitions, as well as unused fields and functions.

A Refactored codebase ensures that code is of the highest quality, with no unused functions or fields, as well as best coding practices.

Anant1902 committed:
Merge pull request #2 from Anant1902/branch-A-CodeQuality

Refactor codebase.

Anant1902 committed:
Add Contact Management

There is no contact management in place.

There is now contact management functionality where a user can add to his contact list and show the contacts in his list.

Anant1902 committed:
Merge pull request #3 from Anant1902/branch-D-Contacts

Add Contact Management

zengzihui commented on src/main/java/drake/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+            fxmlLoader.setRoot(this);
+            fxmlLoader.load();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Perhaps a more user-friendly error message can be provided here to notify users of any errors they encounter rather than printing the stack trace directly to the console.

zengzihui commented on src/main/java/drake/Parser.java:
```diff
@@ -0,0 +1,129 @@
+            date = LocalDate.parse(parts[1], DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            by = date.atStartOfDay();
+        } catch (DateTimeParseException e) {
+            throw new Exception("Date is of the wrong format!");
```
Perhaps you can consider informing users about the accepted date format to enhance clarity and improve the user experience.

zengzihui commented on src/main/java/drake/TodoLeftBlank.java:
```diff
@@ -0,0 +1,12 @@
+ * This is a subclass of StringIndexOutOfBoundsException, which is typically thrown to indicate that
+ * an index is either negative or greater than the size of the string.
+ */
+public class TodoLeftBlank extends StringIndexOutOfBoundsException {
```
A bit of nitpicking but would ToDoLeftBlankException be a better alternative for this exception class name to let it sound like an Exception?

zengzihui reviewed (commented):
generally it looks great! I have learnt a lot from your programming design patterns.

Anant1902 committed:
Improve error handling

Anant1902 committed:
Merge branch 'branch-D-Contacts'

* branch-D-Contacts:
  Improve error handling

Anant1902 committed:
Add contact delete functionality

Anant1902 committed:
Update test classes

Anant1902 committed:
Update Ui.png

Anant1902 committed:
Update docs/README.md

Anant1902 committed:
Update docs/README.md

Anant1902 committed:
Update README.md 3

Anant1902 committed:
Update javadocs.

Anant1902 committed:
Merge commit '7d99e195b0c2493b6ddf477789516b777a614cb1'

* commit '7d99e195b0c2493b6ddf477789516b777a614cb1':
  Update README.md 3
  Update docs/README.md
  Update docs/README.md

Anant1902 committed:
Finalize app

Anant1902 committed:
Refactor Storage.java

Anant1902 committed:
Fix class path issues and finalise bot

Anant1902 committed:
Change state file from .txt to .dat and finalise

Anant1902 committed:
Update README.md
